### PR TITLE
🐛 SSA: ignore diff of other managers

### LINKED ISSUE #6879: SSA dryrun detects changes of other controllers
I played a bit around with SSA and took a closer look at dryrun.

It looks like SSA dryrun currently detects changes made by other controllers and thus triggers SSA even when it's not necessary.

Example:
* Topology Controller: Reconcile: getCurrentState (original)
* KCP controller: takes ownership of additional conditions (managed fields of KCP change)
* Topology Controller: Reconcile: reconcileState (original=>modified)
  * dryrun SSA with modified (picks up changes made by other controllers (i.e. managed field change)
  * diff between original and modified => contains diff made by other controllers (i.e. managedField changes for condition ownership)

I think we should reduce the cases where this can happen to avoid running additional SSA's and potentially even template rotations (although I'm not aware of any infra / bootstrap provider controller which reconciles templates).

I would propose optimizing this by dropping managed fields of other controllers before we diff.

Of course this doesn't catch cases where other controllers change metadata or spec and not only the managed fields.

/kind bug
/area topology
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6879

### CODE REVIEW

**Comment by sbueringer** [comment]
cc @fabriziopandini @chrischdi

File: internal/controllers/topology/cluster/structuredmerge/dryrun.go [hunk 1]
```diff
@@ -157,9 +164,14 @@ func cleanupTopologyDryRunAnnotation(obj *unstructured.Unstructured) error {
 		}
 		managedField.FieldsV1.Raw = fieldsV1Raw
 
-		// Replace the modified managedField entry at the slice.
-		managedFields[i] = managedField
-		obj.SetManagedFields(managedFields)
+		newManagedFields = append(newManagedFields, managedField)
+	}
+
```
Comment (vincepri) [inline internal/controllers/topology/cluster/structuredmerge/dryrun.go:169]: I assume there is a semantic diffence between an empty slice and nil, can we add a comment to clarify what that is and why we need to do it this way?

File: internal/controllers/topology/cluster/structuredmerge/dryrun.go [hunk 2]
```diff
@@ -66,12 +66,12 @@ func dryRunSSAPatch(ctx context.Context, dryRunCtx *dryRunSSAPatchInput) (bool,
 	// and remove the affected managedFields  for `manager=capi-topology` which would
 	// otherwise show the additional field ownership for the annotation we added and
 	// the changed managedField timestamp.
```
Comment (vincepri) [inline internal/controllers/topology/cluster/structuredmerge/dryrun.go:68]: Please add the "why" to these comments

File: internal/controllers/topology/cluster/structuredmerge/dryrun.go (hunk 1 above)
Comment (sbueringer) [reply internal/controllers/topology/cluster/structuredmerge/dryrun.go:169]: Thought about it a bit. I'll drop it here.

In reality managed fields are never empty. There's always something. Adjusted the unit tests so they are closer to reality.

**Review by vincepri — APPROVED** [review]
/lgtm

**Comment by chrischdi** [comment]
Change makes sense to me.

/lgtm

Should we add a proper failing test at `internal/controllers/topology/cluster/structuredmerge/serversidepathhelper_test.go` `TestServerSideApply`:

1. Create obj using topology mgr + ssa
2. modify object using other manager (to get a change in cache so the diff would occur in managedFields when creating the patchhelper)
3. create patchhelper using output obj of 1.
4. assert that there would be no changes?

**Comment by sbueringer** [comment]
> Change makes sense to me.
> 
> /lgtm
> 
> Should we add a proper failing test at `internal/controllers/topology/cluster/structuredmerge/serversidepathhelper_test.go` `TestServerSideApply`:
> 
> 1. Create obj using topology mgr + ssa
> 2. modify object using other manager (to get a change in cache so the diff would occur in managedFields when creating the patchhelper)
> 3. create patchhelper using output obj of 1.
> 4. assert that there would be no changes?

Integrated it one of the tests we already have

**Pull request merged**